src/main/scala/beam/agentsim/infrastructure/ChargingNetwork.scala

```diff
@@ -0,0 +1,246 @@
+      }
+      chargingVehicle
+    }
+
+    /**
+      * remove vehicle from connected list and disconnect from charging point
+      * @param vehicleId vehicle to disconnect
+      * @return status of connection
+      */
+    private[ChargingNetwork] def disconnect(vehicleId: Id[BeamVehicle]): Option[ChargingVehicle] =
+      this.synchronized {
+        connectedVehiclesInternal.remove(vehicleId).map { v =>
+          v.updateStatus(Disconnected)
+          v
+        }
+      }
+
+    /**
+      * process waiting line by removing vehicle from waiting line and adding it to the connected list
+      * @return map of vehicles that got connected
+      */
+    private[ChargingNetwork] def connectFromWaitingLine(): List[ChargingVehicle] = this.synchronized {
+      (1 to Math.min(waitingLineInternal.size, numAvailableChargers)).map { _ =>
+        val v = waitingLineInternal.dequeue()
+        v.updateStatus(Connected)
+        connectedVehiclesInternal.put(v.vehicle.id, v)
+        v
+      }.toList
+    }
+
+    private[ChargingNetwork] def clear(): Unit = this.synchronized {
+      connectedVehiclesInternal.clear()
+      waitingLineInternal.clear()
+    }
+  }
+
+  final case class ChargingCycle(startTime: Int, energy: Double, duration: Long)
+
+  final case class ChargingVehicle(
+    vehicle: BeamVehicle,
+    vehicleManager: VehicleManager,
+    stall: ParkingStall,
+    chargingStation: ChargingStation,
+    arrivalTime: Int,
+    theSender: ActorRef
+  ) extends LazyLogging {
+    import ConnectionStatus._
+    private val chargingSessionInternal = ListBuffer.empty[ChargingCycle]
+    private val statusInternal = ListBuffer.empty[ConnectionStatus]
+
+    private[ChargingNetwork] def updateStatus(status: ConnectionStatus): Unit = this.synchronized {
+      statusInternal.append(status)
+    }
+
+    /**
+      * adding a new charging cycle to the charging session
+      * @param startTime start time of the charging cycle
+      * @param energy energy delivered
+      * @param duration duration of charging
+      * @return boolean value expressing if the charging cycle has been added
+      */
+    def processChargingCycle(startTime: Int, energy: Double, duration: Long): Option[ChargingCycle] =
+      this.synchronized {
+        chargingSessionInternal.lastOption match {
+          case Some(cycle: ChargingCycle) if cycle.startTime == startTime && cycle.duration > duration =>
+            chargingSessionInternal.remove(chargingSessionInternal.size - 1)
+            val cycle = ChargingCycle(startTime, energy, duration)
+            chargingSessionInternal.append(cycle)
+            Some(cycle)
+          case Some(cycle: ChargingCycle) if cycle.startTime == startTime =>
+            // keep existing charging cycle
+            logger.debug(s"the new charging cycle of vehicle $vehicle ends after the end of the last cycle")
+            None
+          case Some(cycle: ChargingCycle) if cycle.startTime > startTime =>
```

Comment:
I would put this as `startTime < cycle.startTime` strictly from a readability perspective. That way it reads as the error points out: "If the new start time is before the stored start time.."

Reply:
yep